### PR TITLE
Add compensations example to Typescript and Java patterns

### LINKED ISSUE #52: Add example with compensations to the examples repo

### PR DESCRIPTION
**Pull request by tillrohrmann**
This PR adds an example for how to use compensations to leave the system in a consistent state in the presence of terminal exceptions (cancellations and deliberate usage of it). The example was added to the TS patterns and the newly created Java patterns. The compensations example is based on [restatedev/restate-holiday](https://github.com/restatedev/restate-holiday).

This fixes #52.

### CODE REVIEW

File: java/README.md [hunk 1]
```diff
@@ -0,0 +1,15 @@
+# Java examples
```
Comment (gvdongen) [inline java/README.md:1]: This feels like repitition of the root README?

File: java/patterns/README.md [hunk 2]
```diff
@@ -0,0 +1,12 @@
...
+A collections of useful patterns in distributed applications, and how to
+implement them with [Restate](https://github.com/restatedev/restate).
+This is a continuously evolving list.
+
+All patterns are described in one file and have a comment block at the top
+that explains them.
+
+### List of Patterns
+
+- [**Compensations**](src/main/java/dev/restate/patterns/Compensations.java)
```
Comment (gvdongen) [inline java/patterns/README.md:12]: Are the patterns meant to be run locally? Or is it more like a reference? Otherwise, it would be nice to have an example invoke cmd to copy over. And a little explanation on how to try out the compensation.

**Review by gvdongen — APPROVED** [review]
LGTM :+1: Will be very helpful for users!

File: java/patterns/src/main/java/dev/restate/patterns/Compensations.java [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+  public static class TravelService extends TravelRestate.TravelRestateImplBase {
+    @Override
+    public void reserve(RestateContext ctx, TravelBookingRequest request) throws TerminalException {
+      final FlightsRestate.FlightsRestateClient flightsService = FlightsRestate.newClient(ctx);
+      final CarRentalRestate.CarRentalRestateClient carRentalService =
+          CarRentalRestate.newClient(ctx);
+      final PaymentRestate.PaymentRestateClient paymentService = PaymentRestate.newClient(ctx);
+
+      // Create a list of compensations to run in case of a failure or cancellation.
+      final List<Runnable> compensations = new ArrayList<>();
```
Comment (slinkydeveloper) [inline java/patterns/src/main/java/dev/restate/patterns/Compensations.java:55]: Just a personal preference for description purposes. I would have used an `ArrayDeque` and then used `descendingIterator()` in the for at the bottom.

File: java/patterns/src/main/resources/log4j2.properties [hunk 4]
```diff
@@ -0,0 +1,12 @@
+# Set to debug or trace if log4j initialization is failing
```
Comment (slinkydeveloper) [inline java/patterns/src/main/resources/log4j2.properties:1]: Please use the new properties from https://github.com/restatedev/sdk-java#logging

File: java/patterns/src/main/java/dev/restate/patterns/Compensations.java [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+   * until it completes. Moreover, an invocation failure and an invocation cancellation are handled
+   * in the exact same way by the caller.
+   */
+  public static class TravelService extends TravelRestate.TravelRestateImplBase {
+    @Override
+    public void reserve(RestateContext ctx, TravelBookingRequest request) throws TerminalException {
+      final FlightsRestate.FlightsRestateClient flightsService = FlightsRestate.newClient(ctx);
+      final CarRentalRestate.CarRentalRestateClient carRentalService =
+          CarRentalRestate.newClient(ctx);
+      final PaymentRestate.PaymentRestateClient paymentService = PaymentRestate.newClient(ctx);
```
Comment (slinkydeveloper) [inline java/patterns/src/main/java/dev/restate/patterns/Compensations.java:52]: Add static imports of the Client classes to reduce the verbosity

File: java/patterns/src/main/java/dev/restate/patterns/Compensations.java [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+
+        final PaymentId paymentId =
+            paymentService
+                .process(PaymentRequest.newBuilder().setTripId(request.getTripID()).build())
+                .await();
+        // Register the compensation to undo the payment.
+        compensations.add(() -> paymentService.refund(paymentId).await());
+
+        flightsService.confirm(flightBookingId).await();
+        carRentalService.confirm(carRentalId).await();
```
Comment (slinkydeveloper) [inline java/patterns/src/main/java/dev/restate/patterns/Compensations.java:80]: Can do a `Awaitable.all()` here.

File: java/patterns/src/main/java/dev/restate/patterns/Compensations.java [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+   * Trip reservation workflow which has been instrumented with compensations. The workflow tries to
+   * reserve the flight and the car rental before it processes the payment. If at any point one of
+   * the calls fails or gets cancelled, then the trip reservation workflow will undo all
+   * successfully completed steps by running the compensations.
+   *
+   * <p>Note: that the compensation logic is purely implemented in the user code and runs durably
+   * until it completes. Moreover, an invocation failure and an invocation cancellation are handled
+   * in the exact same way by the caller.
+   */
+  public static class TravelService extends TravelRestate.TravelRestateImplBase {
```
Comment (slinkydeveloper) [inline java/patterns/src/main/java/dev/restate/patterns/Compensations.java:46]: Perhaps how about adding a test as well? Would it be useful to showcase this scenario even more?

You could add a test that mounts this TravelService, and then mounts a mock FlightService and CarRentalService. The first succeeds, the second fails.

File: java/patterns/src/main/java/dev/restate/patterns/Compensations.java [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+ * completed steps. With Restate, this can be naturally expressed as part of the service code.
+ * Moreover, Restate runs the compensations with the same guarantees as the service code, i.e. it
+ * executes them durably until they complete, which guarantees that the system will be left in a
+ * consistent state.
+ *
+ * <p>The very same mechanism of undoing completed steps via compensations is also needed if you
+ * want to gracefully cancel an invocation, i.e. because it is no longer needed or stuck, and leave
+ * the system in a consistent state. The way a graceful cancellation manifests is by throwing a
+ * `TerminalError` at the next Restate API call. In fact, Restate starts cancelling your invocation
+ * call tree starting with the leave invocations first and then propagating the cancellation error
```
Comment (pcholakov) [inline java/patterns/src/main/java/dev/restate/patterns/Compensations.java:29]: Minor typo! At least I _think_ this is what you mean :-)

```suggestion
 * call tree starting with the leaf invocations first and then propagating the cancellation error
```

File: typescript/patterns/src/compensations.ts [hunk 9]
```diff
@@ -0,0 +1,139 @@
...
+// guarantees as the service code, i.e. it executes them durably until they 
+// complete, which guarantees that the system will be left in a consistent 
+// state.
+//
+// The very same mechanism of undoing completed steps via compensations is 
+// also needed if you want to gracefully cancel an invocation, i.e. because it
+// is no longer needed or stuck, and leave the system in a consistent state.
+// The way a graceful cancellation manifests is by throwing a `TerminalError` 
+// at the next Restate API call. In fact, Restate starts cancelling your
+// invocation call tree starting with the leave invocations first and then
```
Comment (pcholakov) [inline typescript/patterns/src/compensations.ts:25]:
```suggestion
// invocation call tree starting with the leaf invocations first and then 
```

File: java/patterns/src/main/java/dev/restate/patterns/Compensations.java (hunk 3 above)
Comment (pcholakov) [reply java/patterns/src/main/java/dev/restate/patterns/Compensations.java:55]: Good call! I wonder if explicitly reversing a list may not be more obvious what's going on to more readers, though? Then again, the clue is in the `descendingIterator` name!

File: java/README.md (hunk 1 above)
Comment (tillrohrmann) [reply java/README.md:1]: W/o it there is no overview when you browse the java folder (e.g. on Github). I think this is not ideal. Moreover, our current root README is not really hard to parse and read. I think this is something we need to improve.

File: java/patterns/README.md (hunk 2 above)
Comment (tillrohrmann) [reply java/patterns/README.md:12]: The patterns are not meant to be run locally atm. They should compile though. If you think being able to run them adds value, then we can add it as a follow-up.

Comment (gvdongen) [reply java/patterns/README.md:12]: I'm fine with not running them, but maybe we should mention that?

File: java/patterns/src/main/java/dev/restate/patterns/Compensations.java (hunk 3 above)
Comment (tillrohrmann) [reply java/patterns/src/main/java/dev/restate/patterns/Compensations.java:55]: Will update it. I guess either way is fairly straight forward to understand.

File: java/patterns/src/main/resources/log4j2.properties (hunk 4 above)
Comment (tillrohrmann) [reply java/patterns/src/main/resources/log4j2.properties:1]: Will update it.

File: java/patterns/src/main/java/dev/restate/patterns/Compensations.java (hunk 5 above)
Comment (tillrohrmann) [reply java/patterns/src/main/java/dev/restate/patterns/Compensations.java:52]: Will update it.

File: java/patterns/src/main/java/dev/restate/patterns/Compensations.java (hunk 6 above)
Comment (tillrohrmann) [reply java/patterns/src/main/java/dev/restate/patterns/Compensations.java:80]: True.

File: java/patterns/src/main/java/dev/restate/patterns/Compensations.java (hunk 8 above)
Comment (tillrohrmann) [reply java/patterns/src/main/java/dev/restate/patterns/Compensations.java:29]: Thanks for catching this. It is indeed leaf.

File: java/patterns/src/main/java/dev/restate/patterns/Compensations.java (hunk 7 above)
Comment (tillrohrmann) [reply java/patterns/src/main/java/dev/restate/patterns/Compensations.java:46]: Will add a small test.

**Comment by tillrohrmann** [comment]
Thanks a lot for your review @gvdongen, @slinkydeveloper and @pcholakov. I've addressed your comments. Will merge the PR once GHA gives green light.

File: java/patterns/src/test/java/dev/restate/patterns/CompensationsTest.java [hunk 10]
```diff
@@ -0,0 +1,94 @@
...
+        StatusRuntimeException.class,
+        () -> client.reserve(Proto.TravelBookingRequest.newBuilder().setTripID("myTrip").build()),
+        "INTERNAL: Failed to reserve the trip: Failed to reserve car rental. Ran 1 compensations.");
+  }
+
+  // --------------------------------------------------------------------------------------
+  // Helper services for the test
+  // --------------------------------------------------------------------------------------
+
+  private static final class FlightsService extends FlightsRestate.FlightsRestateImplBase {
```
Comment (slinkydeveloper) [inline java/patterns/src/test/java/dev/restate/patterns/CompensationsTest.java:53]: How about `MockFlightsService`? Same for below.
